src/CodeGen_C.cpp

```diff
@@ -1235,9 +1250,19 @@ void CodeGen_C::visit(const Call *op) {
             rhs << "_ucon, ";
         }
 
+        // TODO: special-case for https://github.com/halide/Halide/issues/2099;
+        // embedded buffers will generate code for bounds query, which can
+        // never execute (_halide_buffer_is_bounds_query() always returns false
+        // for them), but the generated bounds-query code won't compile due to 
+        // const correctness. We add some const_casts here as a temporary
+        // workaround.
```

Comment:
That's a nice solution. The type will be mutable, but the shape will be mprotected.

Oh, actually now that I think of it, the buffer *is* mutable, even for embedded buffers. Extern stages might gift it a device allocation or mess with the dirty bits. 

Reading CodeGen_LLVM, it looks like the buffer is always mutable, but the data may or may not be constant. It gets a ConstrantStruct as an initializer, but the GlobalVariable representing the buffer has the isConstant argument unconditionally set to "false".